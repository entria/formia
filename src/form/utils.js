import get from 'lodash.get';
import set from 'lodash.set';
import isEqual from 'lodash.isequal';

export const handleInitialValuesUpdate = (
  currentValues = {},
  currentInitialValues = {},
  newInitialValues = {},
) => {
  const updatedValues = { ...currentValues };

  Object.keys(newInitialValues).map(fieldName => {
    const fieldCurrentValue = get(currentValues, fieldName, null);
    const fieldCurrentInitialValue = get(currentInitialValues, fieldName, null);
    const isFieldPristine = isEqual(fieldCurrentValue, fieldCurrentInitialValue);

    if (!isFieldPristine) {
      return;
    }

    const fieldNewInitialValue = get(newInitialValues, fieldName, null);
    set(updatedValues, fieldName, fieldNewInitialValue);
  });

  return updatedValues;
};
