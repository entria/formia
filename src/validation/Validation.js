import get from 'lodash/get';
import { Strings } from '@entria/utils';

export * as String from './StringValidation';

type Error = {
  code: string,
  message: string,
};

type ValidationValues = {
  [string]: any,
};

type ValidationRules = {
  [string]: any,
};

type ValidationErrors = {
  [string]: Array<Error>,
};

export const validate = (
  values: ValidationValues,
  validations: ValidationRules = {},
): ValidationErrors => {
  const errors = {};

  Object.keys(validations).forEach(field => {
    const value = get(values, field);
    const rules = validations[field];
    if (!rules) {
      return;
    }

    const fieldErrors = rules.map(rule => rule(value)).filter(error => error !== null);

    if (fieldErrors.length > 0) {
      errors[field] = fieldErrors;
    }
  });

  return errors;
};

export const required = (message = 'Required field') => value => {
  const isValid = !Strings.isEmpty(value);
  return isValid ? null : message;
};
