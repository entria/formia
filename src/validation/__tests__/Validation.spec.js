import * as Validation from '../Validation';

import { setupTest } from '../../../test/utils';

beforeEach(async () => setupTest());

it('validate - empty', () => {
  const values = {};
  const validations = {};

  expect(Validation.validate(values, validations)).toMatchSnapshot();
});

it('validate - valid instance', () => {
  const values = { name: 'John Doe' };
  const validations = { name: [Validation.required()] };

  expect(Validation.validate(values, validations)).toMatchSnapshot();
});

it('validate - invalid instance', () => {
  const values = {};
  const validations = { name: [Validation.required()] };

  expect(Validation.validate(values, validations)).toMatchSnapshot();
});

it('required - empty', () => {
  expect(Validation.required()()).toMatchSnapshot();
});

it('required - filled', () => {
  expect(Validation.required()('Anything here')).toMatchSnapshot();
});
