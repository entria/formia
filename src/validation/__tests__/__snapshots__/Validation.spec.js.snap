// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`required - empty 1`] = `
Object {
  "code": "Validation.required",
  "message": "Required field",
  "params": Array [],
}
`;

exports[`required - filled 1`] = `null`;

exports[`validate - empty 1`] = `Object {}`;

exports[`validate - invalid instance 1`] = `
Object {
  "name": Array [
    Object {
      "code": "Validation.required",
      "message": "Required field",
      "params": Array [],
    },
  ],
}
`;

exports[`validate - valid instance 1`] = `Object {}`;
